bootstrap/knownhosts.go

```diff
@@ -33,147 +28,117 @@ func findKnownHosts(sh *shell.Shell) (*knownHosts, error) {
 	sshDirectory := filepath.Join(userHomePath, ".ssh")
 	knownHostPath := filepath.Join(sshDirectory, "known_hosts")
 
-	// Ensure a directory exists and known_host file exist
-	if !fileExists(knownHostPath) {
-		os.MkdirAll(sshDirectory, 0755)
-		ioutil.WriteFile(knownHostPath, []byte(""), 0644)
+	// Ensure ssh directory exists
+	if err := os.MkdirAll(sshDirectory, 0700); err != nil {
+		return nil, err
 	}
 
-	lockFile := filepath.Join(sshDirectory, "known_hosts.lock")
-
-	// Create a lock on the known_host file so other agents don't try and
-	// change it at the same time
-	knownHostLock, err := shell.LockFileWithTimeout(sh, lockFile, time.Second*30)
-	if err != nil {
-		return nil, fmt.Errorf("Could not acquire a lock on `%s`: %v", lockFile, err)
+	// Ensure file exists
+	if _, err := os.Stat(knownHostPath); err != nil {
+		f, err := os.OpenFile(knownHostPath, os.O_CREATE|os.O_WRONLY, 0600)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Could not create %q", knownHostPath)
+		}
+		if err = f.Close(); err != nil {
+			return nil, err
+		}
 	}
 
-	return &knownHosts{knownHostLock, sh, knownHostPath}, nil
+	return &knownHosts{Shell: sh, Path: knownHostPath}, nil
 }
 
 func (kh *knownHosts) Contains(host string) (bool, error) {
-	sshToolsDir, err := findSSHToolsDir(kh.shell)
+	file, err := os.Open(kh.Path)
 	if err != nil {
 		return false, err
 	}
-
-	// Grab the generated keys for the repo host
-	keygenOutput, err := kh.shell.RunAndCapture(
-		filepath.Join(sshToolsDir, "ssh-keygen"), "-f", kh.Path, "-F", host)
-
-	// Returns an error and no output if host isn't in there
-	if err != nil && keygenOutput == "" {
-		return false, nil
-	} else if err != nil {
-		return false, fmt.Errorf("Could not perform `ssh-keygen` (%s)", err)
+	defer file.Close()
+
+	normalized := knownhosts.Normalize(host)
+
+	// There don't appear to be any libraries to parse known_hosts that don't also want to
+	// validate the IP's and host keys. Shelling out to ssh-keygen doesn't support custom ports
+	// so I guess we'll do it ourselves.
+	//
+	// known_host format is defined at https://man.openbsd.org/sshd#SSH_KNOWN_HOSTS_FILE_FORMAT
+	// A basic example is:
+	// # Comments allowed at start of line
+	// closenet,...,192.0.2.53 1024 37 159...93 closenet.example.net
+	// cvs.example.net,192.0.2.10 ssh-rsa AAAA1234.....=
+	// # A hashed hostname
+	// |1|JfKTdBh7rNbXkVAQCRp4OQoPfmI=|USECr3SWf1JUPsms5AqfD5QfxkM= ssh-rsa
+	// AAAA1234.....=
+	// # A revoked key
+	// @revoked * ssh-rsa AAAAB5W...
+	// # A CA key, accepted for any host in *.mydomain.com or *.mydomain.org
+	// @cert-authority *.mydomain.org,*.mydomain.com ssh-rsa AAAAB5W...
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		fields := strings.Split(scanner.Text(), " ")
+		if len(fields) != 3 {
+			continue
+		}
+		for _, addr := range strings.Split(fields[0], ",") {
+			if addr == normalized {
```

Comment:
Do we are that this no longer supports hashed hosts?

Reply:
Yeah, good point. I'll add that.